### PR TITLE
[WIP] [Issue#4939] Changes to support new purge task executions and purge workflows

### LINKED ISSUE #4939: Support purge operations for st2 Workflows and TaskExecution
## SUMMARY

We are looking for options and potential solutions on an issue in stack storm stack where 2 collections with-in mongodb bloat quite a bit - st2.task_execution_d_b, st2.workflow_execution_d_b. 

We have a cron job to purge old data from mongo which uses st2-purge-executions, st2-purge-trigger-instances, curious if we have any flags or scripts to purge st2.task_execution_d_b and st2.workflow_execution_d_b collections(tables) explicitly ?

Our current approach to the customer is to drop the entire st2 db which is not ideal. 

I looked at the current st2 purge operations and took a similar approach here is the PR : #4924 , please let me know if the approach makes sense.

### STACKSTORM VERSION

st2 --version :  st2 3.1.0, on Python 2.7.15

##### OS, environment, install method
A Centos VM : centos-release-7-8.2003.0.el7.centos.x86_64

Install method: Custom install - rpm.

## Steps to reproduce the problem

The issue is caused with many workflows running over time and bloating the mongoDB. Here is the snapshot of the mongoDB

Here's a snapshot of the collections ( format : <collection_name>: <size> <units> ( <storageSize> <units>) )

st2.task_execution_d_b: 10.2 GB (1.9 GB)
st2.workflow_execution_d_b: 10.2 GB (1.8 GB)
st2.trace_d_b: 1.1 GB (334.0 MB)
st2.action_execution_d_b: 752.7 MB (165.2 MB)
st2.trigger_instance_d_b: 426.1 MB (107.1 MB)
st2.live_action_d_b: 292.4 MB (72.8 MB)
st2.rule_enforcement_d_b: 279.1 MB (82.5 MB)
st2.token_d_b: 219.1 MB (65.4 MB)
st2.action_execution_output_d_b: 18.8 MB (5.2 MB)
st2.action_d_b: 99.6 kB (48.0 kB)
st2.rule_d_b: 44.9 kB (28.0 kB)
st2.runner_type_d_b: 33.3 kB (24.0 kB)
st2.pack_d_b: 23.9 kB (20.0 kB)
st2.trigger_type_d_b: 16.1 kB (20.0 kB)
st2.sensor_type_d_b: 7.3 kB (16.0 kB)

## Expected Results

What did you expect to happen when running the steps above?
Expect a support for purging workflow executions  and task executions entries in the mongo DB.

## Actual Results

What happened? What output did you get?

Making sure to follow these steps will guarantee the quickest resolution possible.

Thanks!

### PR DESCRIPTION
**Pull request by srimandaleeka01**
Closes #4939

### CODE REVIEW

**Comment by srimandaleeka01** [comment]
> @srimandaleeka01 I think overall these changes are great to see.
> 
> What i think would be a better approach though is to integrate the workflow and task purging into the already existing purge executions script. This way things like the `st2garbagecollector` process will be automatically updated.
> 
> Should be just moving some code around.
> 
> Does that sound good?

Thanks for the pointer @nmaludy . Sorry have been out of loop on this project. I made the changes to call purge workflow and task executions in the garbagecollector . Please take a look : https://github.com/StackStorm/st2/pull/4924/commits/36c7fdf50b6675df4c1fd030cc08661e196ef551 

With the above change I also added 2 new attributes `_workflow_execution_ttl` `_task_execution_ttl` this is causing the unit tests to fail . How can I fix the unit test ? Couldn't figure out yet where the st2.conf file is being patched or where the fixture is 

```1) ERROR: test_process_error_handling_has_error (tests.unit.test_workflow_engine.WorkflowExecutionHandlerTest)
----------------------------------------------------------------------
   Traceback (most recent call last):
    /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/mock/mock.py line 1305 in patched
      return func(*args, **keywargs)
    st2actions/tests/unit/test_workflow_engine.py line 226 in test_process_error_handling_has_error
      gc = garbage_collector.GarbageCollectorService()
    st2reactor/st2reactor/garbage_collector/base.py line 72 in __init__
      self._workflow_execution_ttl = cfg.CONF.garbagecollector.workflow_execution_ttl
    /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/oslo_config/cfg.py line 2547 in __getattr__
      return self._conf._get(name, self._group)
    /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/oslo_config/cfg.py line 2264 in _get
      value = self._do_get(name, group, namespace)
    /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/oslo_config/cfg.py line 2282 in _do_get
      info = self._get_opt_info(name, group)
    /opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/oslo_config/cfg.py line 2415 in _get_opt_info
      raise NoSuchOptError(opt_name, group)
   NoSuchOptError: no such option in group garbagecollector: workflow_execution_ttl```


Thanks again !

**Comment by cognifloyd** [comment]
We just reformatted the code with `black`. So, would you merge master into your branch (resolve any conflicts & reformat with black)? This should help resolve any linting errors. Thanks!

**Comment by copart-jafloyd** [comment]
@srimandaleeka01 I'd really like to see this in v3.6.0 if possible. Do you have bandwidth to bring this up-to-date with master?

Both unit and lint failures are due to missing config options. You can add those here:
https://github.com/StackStorm/st2/blob/2a6ec4f68f849b2b9289af80c3a3031a031bfba5/st2reactor/st2reactor/garbage_collector/config.py#L81-L99

**Comment by cognifloyd** [comment]
@srimandaleeka01 I'd really like to see this in v3.6.0 if possible. Do you have bandwidth to bring this up-to-date with master?

Both unit and lint failures are due to missing config options. You can add those here:
https://github.com/StackStorm/st2/blob/2a6ec4f68f849b2b9289af80c3a3031a031bfba5/st2reactor/st2reactor/garbage_collector/config.py#L81-L99

Sorry for the duplicate message. I was logged into the wrong github account. I just deleted the other message.

**Comment by cognifloyd** [comment]
Also, could you edit your PR description to include `Closes: #4939`? Thanks!

**Comment by cognifloyd** [comment]
@srimandaleeka01 Would you be willing to allow maintainers to edit your PR? If so, I can quickly merge in master and adjust the config options for you.

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (amanda11) [inline st2reactor/st2reactor/garbage_collector/base.py:257]: Think this should be _task_execution_ttl instead of actions

**Review by amanda11 — CHANGES_REQUESTED** [review]

**Comment by srimandaleeka01** [comment]
> @srimandaleeka01 Would you be willing to allow maintainers to edit your PR? If so, I can quickly merge in master and adjust the config options for you.

Please go ahead and the maintainers can edit the PR for getting it merged. Thank you!

**Comment by amanda11** [comment]
@cognifloyd  I can update the PR, as I have taken this code made some slight changes and got it installed on a ST2 system. So happy to do update to master and merge those changes in.

**Comment by amanda11** [comment]
@srimandaleeka01 Could you check if maintainers have access to this PR request, as I'm getting 403 error trying to push. 

Please see: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork for info on how to allow maintainers to edit a PR that has been created from a fork.

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (srimandaleeka01) [reply st2reactor/st2reactor/garbage_collector/base.py:233]: Thank you @amanda11 I will make this change.

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (srimandaleeka01) [reply st2reactor/st2reactor/garbage_collector/base.py:257]: > Think this should be _task_execution_ttl instead of actions

Thank you @amanda11 will make this change.

File: st2common/st2common/cmd/purge_task_executions.py
Comment (amanda11) [inline st2common/st2common/cmd/purge_task_executions.py:1]: Copyright is now StackStorm Authors, and year file first introduced, e.g. could you change to:

# Copyright 2021, The StackStorm Authors.

File: st2common/st2common/cmd/purge_workflows.py
Comment (amanda11) [inline st2common/st2common/cmd/purge_workflows.py:1]: # Copyright 2021, The StackStorm Authors.

File: st2common/st2common/garbage_collection/workflows.py
Comment (amanda11) [inline st2common/st2common/garbage_collection/workflows.py:1]: # Copyright 2021, The StackStorm Authors.

File: st2common/tests/unit/test_purge_task_executions.py
Comment (amanda11) [inline st2common/tests/unit/test_purge_task_executions.py:1]: # Copyright 2021, The StackStorm Authors.

**Comment by srimandaleeka01** [comment]
> @srimandaleeka01 Could you check if maintainers have access to this PR request, as I'm getting 403 error trying to push.
> 
> Please see: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork for info on how to allow maintainers to edit a PR that has been created from a fork.

I was following the article above to allow mainters to edit the PR, however I don't see that option. 
It's likely that the PR has been created from a Plexxi:st2 ( a fork from master:st2). 

Are there any specific changes apart from the comments you have posted for the PR ? If so please post the diff for the corresponding file , I can manually apply.

File: st2common/tests/unit/test_purge_worklows.py
Comment (amanda11) [inline st2common/tests/unit/test_purge_worklows.py:1]: # Copyright 2021, The StackStorm Authors.

File: st2common/tests/unit/test_purge_task_executions.py
Comment (amanda11) [inline st2common/tests/unit/test_purge_task_executions.py:19]: The import of TRIGGER_INSTANCE_PROCESSED isn't used, so it breaks the CI checks.

File: st2common/tests/unit/test_purge_worklows.py
Comment (amanda11) [inline st2common/tests/unit/test_purge_worklows.py:19]: As TRIGGER_INSTANCE_PROCESSED is unused it will break the CI checks, so can you remove this import.

File: st2common/st2common/garbage_collection/workflows.py
Comment (amanda11) [inline st2common/st2common/garbage_collection/workflows.py:71]: At the moment the action_ref passed in from the standalone tool, doesn't take affect - as its not used in this class. I don't believe action_ref is valid for WorkflowExecution or TaskExecution, so I would remove the commented out lines and also the action_ref in the cmd/purge_....py classes.

**Review by amanda11 — CHANGES_REQUESTED** [review]
Thanks so much for this PR. It will be really great to get this into the next release of StackStorm.

Just some minor other comments I found, in addition to fix the CircleCI then:
1. Need to run black on the files changed, so that it doesn't break the lint checks
2. Need to add the change to st2reactor/st2reactor/garbage_collector/config.py as mentioned earlier in review comments to prevent the UT failure to add the new config options.

Can you also update CHANGELOG.rst - to update the "Added" section to include details of this PR?

If need any assistance or want us to help with doing any of the amendments then just please let us know (we'd need the fork updated with permissions for maintainers to edit as I think at moment it's blocked). But if you are happy to do the changes there would be no need.

**Comment by amanda11** [comment]
> > @srimandaleeka01 Could you check if maintainers have access to this PR request, as I'm getting 403 error trying to push.
> > Please see: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork for info on how to allow maintainers to edit a PR that has been created from a fork.
> 
> I was following the article above to allow mainters to edit the PR, however I don't see that option. It's likely that the PR has been created from a Plexxi:st2 ( a fork from master:st2) which is more restrictive ( intentionally ) .
> 
> Are there any specific changes apart from the comments you have posted for the PR ? If so please post the diff for the corresponding file , I can manually apply.

Hi - I've listed in that last review I think all the changes that I was going to make - just a few minor things like copyright, black formatting etc, and then I had a look at the TODO you had left around action_ref, and checked the collection contents - so think safest to remove the action_ref that was in the TODO whether it was relevant or not.
There is also a merge conflict with master - but the instructions to fix that are provided by GIT where it says there are conflicts (it's just that in setup.py on master it has a new binary in its list, so just a small conflict on that to change).

**Comment by amanda11** [comment]
@srimandaleeka01 Are you ok to sign the CLA Contribution Agreement? At the moment it is blocking the PR as the contribution agreement isn't signed.
We really appreciate this fix, and we'd like to get it into the 3.7.0 release, which is targetted for end of March.

Do you have time to address the review comments this month, as we'd like to get this into the 3.7.0 PR. Alternatively, if the contribution agreement is signed, then we could fork off this repo and then merge a PR that way - and ask one of the TSC to make the review comments? Whichever you prefer.

Thanks once again for the contribution.

**Pull request closed without merging**
